### ISSUE TITLE
Using functions as keys returns stale data

### ISSUE THREAD

**Issue by ath92** [vercel/swr#611]
# Bug report

I'm using SWR to build an app that has an existing "client" layer. Imagine something like the following syntax for this client layer, which returns a Promise (i.e. calling this function does the actual fetching):
```
client.account(userID).getAvatar();
```

It's important to note that the identity of `getAvatar` in this example is stable as long as `userID` doesn't change, i.e.
```
const one = client.account(userID).getAvatar;
const two = client.account(userID).getAvatar;
one === two // this is true, because getAvatar has a stable identity
```

## Description / Observed Behavior

Because of this stable identity, I want to be able to use these functions directly as part of the cache key in swr, like so:
```
const { getAvatar } = client.account(userID);
const { data } = useSWR([getAvatar]);
```

Where `userID` is provided from higher up, e.g. through some sort of context. `getAvatar` in this case knows about `userID` because it has it in its closure.

Unfortunately, it turns out swr will not update `data` if the function is changed, as a result of `userID` changing.

## Expected Behavior

I'd expect swr to be able to use function as keys. I'm okay with wrapping these functions in an array, like I did in the example above.

## Repro Steps / Code Example

[CodeSandbox repro](https://codesandbox.io/s/swr-function-keys-4mjyb?file=/src/App.js)

## Additional Context
I had a quick look through the source code and I believe the issue is here: https://github.com/vercel/swr/blob/daf7eadf4fa92913daf19d093a3dd26759d503da/src/libs/hash.ts#L11
Specifically, functions are stringified and added to the hash, but since `userID` in the example above is captured in is closure, the string output of that operation doesn't change when `userID` changes. 

Using version 0.3.0

**Comment by shuding** [comment]
Hello @ath92! Currently, what you can do is to use `userID` as the key and `getAvatar` as the fetcher:

```js
const { data } = useSWR(userID, client.account(userID).getAvatar)
```

And then simplify it as a custom hook, for example:
```js
const useAvatar = userID => useSWR(userID, client.account(userID).getAvatar)

// ...
const { data } = useAvatar(userID)
```

What do you think about it?

**Comment by ath92** [comment]
Hey, this would work for the simple case. But as far as I understand, the key (`userID` in the example you provided) is used as an identifier in the global cache, correct?

So if I understand correctly, it would mean that the following:
```
const { data } = useSWR(userID, client.account(userID).getAvatar)
```

would be given the same key (and therefore use the same cached values) as this (note `getAvatar` -> `getUsername`):
```
const { data } = useSWR(userID, client.account(userID).getUsername)
```

because only `userID` is provided as a key here, and the fetcher function is not used as a key for the global cache. I'm not 100% sure if this is correct, but maybe you can verify whether this is in fact the case?

---

Additionally, the example you provided works well enough for a simple use case, but sometimes I'd like to be able to pass more parameters to the fetcher. Imagine for example the following case:

```
// this is client.account(userID).getAvatar
getAvatar(width = 0, height = 0) {
  return client.get(`someUrl/avatar?width=${width}&height=${height});
}
```

Then I'd like to be able to do the following:
```
const { getAvatar } = client.account(userID);
const { data } = useSWR([getAvatar, 40, 40]);
```

I can do something like this by providing a single, simple fetcher to the global SWR config that looks like this:
```
const simpleFetcher = (resourceMethod, ...args) => resourceMethod(...args);
```

This example wouldn't work with the workaround you mentioned, because in that case `userID` will get passed as the first argument to the resource method.

**Comment by ath92** [comment]
After thinking about this a little bit more, I think it might also make sense to let the `fetcher` function be part of the cache key as well. 

This would enable a syntax like the following:
```javascript
const fetcher = client.account(userID).getAvatar;

const Comp = () => {
   const { data } = useSWR(['some param'], fetcher);
  // ...
}
```

This approach doesn't work (I think) in the current implementation, because if I were to use the same `some param` with another fetcher, I would get the same cached results, even though the fetcher could be doing completely different stuff.

**Comment by ath92** [comment]
@shuding Do you have an update on this? The PR I opened earlier would allow me to use `swr` the way I would like to. Is there a reason why it shouldn't be merged?

I'm also happy to try and help by finding a different approach/fix, but I'd like to know if it makes sense to do that.

**Comment by shuding** [comment]
> Hey, this would work for the simple case. But as far as I understand, the key (`userID` in the example you provided) is used as an identifier in the global cache, correct?

Yeah, the key should be the identifier of your data. Personally, I will add more context to the key so it won't conflict:

```js
useSWR(['getAvatar', userID], client.account(userID).getAvatar)
useSWR(['getUsername', userID], client.account(userID).getUsername)
```

And then, it's better to pass parameters to the `fetcher`:

```js
useSWR(['getAvatar', userID], (method, id) => client.account(id)[method](id))
useSWR(['getUsername', userID], (method, id) => client.account(id)[method](id))
```

Furthermore, I can use a global fetcher, don't even need to pass it anywhere:

```js
const fetcher = (method, id) => client.account(id)[method](id)

// ...
useSWR(['getAvatar', userID], fetcher)
useSWR(['getUsername', userID], fetcher)
```

You can have more flexibility to pass extra params too if you define `fetcher` like this:
```js
const fetcher = (method, id, ...params) => client.account(id)[method](...params)
```

I think that's simpler. 

However, I see the benefit to allow users to pass functions as keys:

```js
useSWR([getAvatar, userID], (fn, ...args) => fn(...args))
```

I personally don't think it's as straightforward as `useSWR([...args], fn)`, but it solves the key conflict issue indeed.

I'll approve and merge your PR, nice work! @ath92

**Comment by shuding** [comment]
Released your PR with `swr@0.3.2`, thanks again!

**Comment by ath92** [comment]
Awesome, thanks for your help @shuding. I realise my use case is probably not very common, so I appreciate you taking the time to provide examples.
